logstash-core/src/main/java/org/logstash/instrument/metrics/HistoricMetric.java

```diff
@@ -0,0 +1,233 @@
+    private final AtomicReference<Node> quickOneMinute = new AtomicReference<>();
+    private final AtomicReference<Node> quickFiveMinutes = new AtomicReference<>();
+    private final AtomicReference<Node> quickFifteenMinutes = new AtomicReference<>();
+
+    // Safeguard reads against dead writer
+    private final AtomicReference<Instant> lastRecord = new AtomicReference<>();
+
+    public HistoricMetric(Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this(Clock.systemUTC(), numeratorMetric, denominatorMetric);
+    }
+
+    public HistoricMetric(Metric<T> metric) {
+        this(Clock.systemUTC(), metric);
+    }
+
+    HistoricMetric(Clock clock, Metric<T> numeratorMetric, Metric<Long> denominatorMetric) {
+        this.clock = clock;
+        this.numeratorMetric = numeratorMetric;
+        this.denominatorMetric = denominatorMetric;
+
+        // TODO: refactor initial capture out of constructor?
+        this.origin = new Node(clock.instant(), numeratorMetric.getValue(), denominatorMetric.getValue());
+        this.lastRecord.set(this.origin.timestamp);
+        this.newest = new AtomicReference<>(origin);
+    }
+
+    HistoricMetric(Clock clock, Metric<T> metric) {
+        this(clock, metric, new UptimeMetric(clock));
+    }
+
+    // TODO: recording of an HistoricMetric will need to be scheduled
+    public void record() {
+        record(clock.instant(), numeratorMetric.getValue(), denominatorMetric.getValue());
+    }
+
+    void record(final Instant timestamp, final T numerator, final Long denominator) {
+        this.newest.getAndUpdate((previousNewest) -> previousNewest.seekAndPrepend(timestamp, numerator, denominator).getNewest());
+        this.lastRecord.set(timestamp);
+
+        // rotate our quick references in ascending age order, bailing entirely if we are too young
+        if (!rotate(quickInstant, timestamp.minusSeconds(1))) { return; }
+        if (!rotate(quickOneMinute, timestamp.minusSeconds(60))) { return; }
+        if (!rotate(quickFiveMinutes, timestamp.minusSeconds(300))) { return; }
+        if (!rotate(quickFifteenMinutes, timestamp.minusSeconds(900))) { return; }
+
+        // trim out elements that are no longer necessary
+        compact();
+    }
+
+    // returns true IFF our rotation ended up with a non-null rotation reference
+    private boolean rotate(final AtomicReference<Node> rotationReference, final Instant threshold) {
+        Node result = rotationReference.updateAndGet((ref) -> Objects.requireNonNullElse(ref, origin).seekAhead(threshold));
+        return Objects.nonNull(result);
+    }
+
+    private void compact() {
+        final Node quick_15 = quickFifteenMinutes.get();
+        if (quick_15 != null) {
+            origin.nextNode.set(quick_15);
+        }
+    }
+
+    public Map<String,Double> availableRates() {
+        if (lastRecord.get().isBefore(clock.instant().minusSeconds(10))) {
```

Comment:
curious why we are capturing the record every 10 sec?

Reply:
This still needs to be hooked up to en executor service to capture metrics on a regular cadence, but the thought here was that on read, a stale metric that was for some reason not being updated it would be better to limit just how stale it was allowed to get.